### PR TITLE
[SPARK-29116][PYTHON][ML] Refactor py classes related to DecisionTree

### PR DESCRIPTION
**Pull request by huaxingao**
### What changes were proposed in this pull request?

- Move tree related classes to a separate file ```tree.py```
- add method ```predictLeaf``` in ```DecisionTreeModel```& ```TreeEnsembleModel```

### Why are the changes needed?

- keep parity between scala and python
- easy code maintenance

### Does this PR introduce any user-facing change?
Yes
add method ```predictLeaf``` in ```DecisionTreeModel```& ```TreeEnsembleModel```
add ```setMinWeightFractionPerNode```  in ```DecisionTreeClassifier``` and ```DecisionTreeRegressor```

### How was this patch tested?
add some doc tests

### CODE REVIEW

**Comment by zhengruifeng** [comment]
retest this please

File: python/pyspark/ml/regression.py [hunk 1]
```diff
@@ -1252,9 +999,29 @@ def featureImportances(self):
         return self._call_java("featureImportances")
 
 
+class GBTRegressorParams(GBTParams, TreeRegressorParams):
```
Comment (zhengruifeng) [inline python/pyspark/ml/regression.py]: What about rename it `_GBT...`? like above `_RandomForestClassifierParams`

File: python/pyspark/ml/tree.py [hunk 2]
```diff
@@ -0,0 +1,348 @@
...
+        """
+        Predict the indices of the leaves corresponding to the feature vector.
+        """
+        return self._call_java("predictLeaf", value)
+
+    def __repr__(self):
+        return self._call_java("toString")
+
+
+class DecisionTreeParams(HasCheckpointInterval, HasSeed, HasWeightCol):
```
Comment (zhengruifeng) [inline python/pyspark/ml/tree.py]: ditto

File: python/pyspark/ml/tree.py [hunk 3]
```diff
@@ -0,0 +1,348 @@
...
+from pyspark import since, keyword_only
+from pyspark.ml.param.shared import *
+from pyspark.ml.util import *
+from pyspark.ml.wrapper import JavaEstimator, JavaModel, JavaParams, \
+    JavaPredictor, JavaPredictionModel
+from pyspark.ml.common import inherit_doc, _java2py, _py2java
+
+
+@inherit_doc
+class DecisionTreeModel(JavaPredictionModel):
```
Comment (zhengruifeng) [inline python/pyspark/ml/tree.py]: ditto

File: python/pyspark/ml/tree.py [hunk 4]
```diff
@@ -0,0 +1,348 @@
...
+
+    def getCacheNodeIds(self):
+        """
+        Gets the value of cacheNodeIds or its default value.
+        """
+        return self.getOrDefault(self.cacheNodeIds)
+
+
+@inherit_doc
+class TreeEnsembleModel(JavaPredictionModel):
```
Comment (zhengruifeng) [inline python/pyspark/ml/tree.py]: ditto

**Comment by huaxingao** [comment]
@zhengruifeng 
Thanks for your comments.
I didn't add _single_leading_underscore for classes that are used for other packages. 
I am a little fuzzy about this _single_leading_underscore usage:
In https://pep8.org/#descriptive-naming-styles, it has
```_single_leading_underscore: weak “internal use” indicator. E.g. from M import * does not import objects whose name starts with an underscore.```
It makes me feel that the class with _single_leading_underscore is for internal use only. It is not intended to be used in other packages. However, if I explicitly import the _single_leading_underscore class, it works OK. 
For example, If I do 
```from pyspark.ml.tree import *```, the _single_leading_underscore class is not imported. 
If I do
```from pyspark.ml.tree import _DecisionTreeModel, _DecisionTreeParams```, these classes are imported OK.

**Comment by zhengruifeng** [comment]
@huaxingao Yes, I can reproduce your case.
The 'private' classes can only be imported explicitly. I guess that is way it is a **weak** “internal use” indicator.
I think we can add _single_leading_underscore according to the scala side.

**Comment by huaxingao** [comment]
OK. I will add _single_leading_underscore to the classes you mentioned in the comments. Thanks!

**Comment by zhengruifeng** [comment]
retest this please

**Comment by zhengruifeng** [comment]
retest this please

**Review by zhengruifeng — APPROVED** [review]

**Comment by zhengruifeng** [comment]
Merged to master, thanks @huaxingao

**Comment by huaxingao** [comment]
Thanks a lot for your help! @zhengruifeng

**Pull request closed without merging**